tv_apps/smart-home/js/filter_manager.js

```diff
@@ -206,27 +206,54 @@
       if (!this._isFilterChanging) {
         return;
       }
-
-      var that = this;
       var filter = this.getFilterByIconName(this._cardFilter.filter);
-      var gotCardLists = function(filteredList) {
-        filteredList.forEach(function(card) {
-          that._cardScrollable.addNode(that._home.createCardNode(card));
-        });
-
-        that._filteredCardList =
-          (filter.name !== 'all') ? filteredList : undefined;
-        that._cardListElem.style.opacity = 0;
-        window.requestAnimationFrame(that._performBubbleUp.bind(that));
-      };
 
       if (!this._isBubbleSinking()) {
         this._isFilterChanging = false;
         return;
       }
 
+      this._filteredCardList = undefined;
+
+      var display = list => {
+        list.forEach(card => {
+          this._cardScrollable.addNode(this._home.createCardNode(card));
+        });
+        this._cardListElem.style.opacity = 0;
+        window.requestAnimationFrame(this._performBubbleUp.bind(this));
+      };
+
       this._cardScrollable.clean();
-      this._cardManager.getFilteredCardList(filter.name).then(gotCardLists);
+      if (filter.name === 'all') {
+        // If filtering condition is all cards, just display all cards.
+        // Don't need extra card sorting.
+        this._cardManager.getFilteredCardList(filter.name).then(display);
+      } else {
+        this._cardManager.getFilteredCardList(filter.name).then(list => {
+          var appearedList = [];
+          var uniqueList = [];
+          list.forEach(card => {
+            // Get url to which card is launched
+            var url = card.launchURL ? card.launchURL : card.manifest.origin;
+            // Get sorting key based on card name
+            var key = CardUtil.getSortKey(card);
+            // The same card which appeared before will not be picked up.
+            // By the same card, we mean card has the same url and name.
+            // (the same name would generate the same key so we use key
+            // for checking here)
+            if (appearedList.indexOf(url + key) < 0) {
+              appearedList.push(url + key);
+              uniqueList.push({
+                card: card,
+                key: key
+              });
+            }
+          });
+          uniqueList.sort((a, b) => a.key.localeCompare(b.key));
```

Comment:
I would recommend you to remove duplicate cards **after** sorting for better performance. (although it doesn't really matter with that small scale, it's worth a try.)

Like this:

```js
var lastItem;
list.reduce(function(reducedArray, item) {
  if (item !== lastItem) {
    lastItem = item;
    reducedArray.push(item);
  }
  return reducedArray;
}, []);
```